src/utils/streams/concat_stream_providers.js

```diff
@@ -0,0 +1,42 @@
+import { PassThrough } from 'stream';
+
+/**
+ *  Write the data and errors from a list of stream providers
+ *  to a single stream in order. Stream providers are only
+ *  called right before they will be consumed, and only one
+ *  provider will be active at a time.
+ *
+ *  @param {Array<() => ReadableStream>} sourceProviders
+ *  @param {PassThroughOptions} options options passed to the PassThrough constructor
+ *  @return {WritableStream} combined stream
+ */
+export function concatStreamProviders(sourceProviders, options = {}) {
+  const destination = new PassThrough(options);
+  const queue = sourceProviders.slice();
+
+  (function pipeNext() {
+    const provider = queue.shift();
+
+    if (!provider) {
```

Comment:
__question:__ isn't this case already covered by `if (!isLast) { source.once('end', pipeNext); }`? Or you expect `falsy` elements in `sourceProviders` list or just that this list could be empty (that, it seems, will cause `destination` stream to never "end")?

Reply:
Yeah, but shouldn't we throw or end `destination` stream immediately in this case?